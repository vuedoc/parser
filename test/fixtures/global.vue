<template>
  <div>
    <!-- Template event with @ -->
    <input @input="$emit('template-@-event', $event)"/>

    <!-- Template event with v-on -->
    <input v-on:input="$emit('template-v-on-event', $event)" />

    <label>
      <input :disabled="disabled" type="text" v-model="checkbox"/>
      <!-- Default slot -->
      <slot></slot>
      <!-- Use this slot to set the checkbox label -->
      <slot name="label">Unamed checkbox</slot>
      <!--
        This
        is multiline description
      -->
      <slot name="multiline">Unamed checkbox</slot>
      <slot name="undescribed"></slot>
      <!--
        Slot with defined keywords

        @protected
        @props { name, address, email }
      -->
      <slot name="multiline">Unamed checkbox</slot>
      <template></template>
    </label>
  </div>
</template>

<script>
  const componentName = 'checkboxPointer'
  const aliasName = componentName

  /**
   * The generic component
   * Sub description
   *
   *
   * @public
   * @alpnum azert0 123456789
   * @generic Keyword generic description
   * @multiline Keyword multiline
   *            description
   * @special-char {$[ç(àë£€%µù!,|`_\\<>/_ç^?;.:/!§)]}
   * @punctuations !,?;.:!
   * @operators -/+<>=*%
   *
   * @slot inputs - Use this slot to define form inputs ontrols
   * @slot actions - Use this slot to define form action buttons controls
   * @slot footer - Use this slot to define form footer content.
   */
  export default Vue.extend({
    name: 'checkbox',
    ['name']: componentName,
    model: {
      prop: 'model',
    },
    props: {
      /**
       * The checkbox model
       */
      model: {
        type: Array,
        required: true
      },

      /**
       * Initial checkbox state
       */
      disabled: Boolean,

      /**
       * Initial checkbox value
       */
      checked: {
        type: Boolean,
        default: true
      },

      // Prop with arrow function
      propWithArrow: {
        type: Object,
        default: () => ({ name: 'X'})
      },

      // Prop with object function
      propWithFunction: {
        type: Object,
        default() {
          return { name: 'X' }
        }
      },

      // Prop with anonymous function
      propWithAnonymousFunction: {
        type: Object,
        default: function() {
          return { name: 'X' }
        }
      },

      // Prop with named function
      propWithNamedFunction: {
        type: Object,
        default: function propWithNamedFunction() {
          return { name: 'X' }
        }
      },

      /**
       * Prop with protected visibility
       *
       * @private
       */
      propWithPrivateVisibility: Boolean
    },

    data () {
      const pointer = 'pointed value'

      return {
        int: 12,
        float: 12.2,
        booleanTrue: true,
        booleanFalse: false,
        string: 'Hello',
        string: 'Hello',
        null: null,
        undefined: undefined,
        function: new Function(),
        arrowFunction: () => undefined,
        date: Date.now(),
        pointer: pointer,
        componentName: componentName,
        value: null
      }
    },

    data: {
      int: 13
    },

    computed: {
      id () {
        const name = this.componentName

        return `${name}-${this.pointer}`
      },
      name () {
        return this.componentName
      },
      withoutDependency () {
        return 12
      }
    },

    methods: {
      /**
       * @private
       */
      privateMethod () {
        console.log('check')

        const name = 'check'
        const value = 'event value'

        if (name) {
          console.log('>', name)
        }

        /**
        * Event with identifier name
        */
        this.$emit(name, value)
      },

      /**
      * Check the checkbox
      */
      check () {
        console.log('check')

        let eventName = 'check'
        const value = 'event value'

        if (eventName) {
          console.log('>', eventName)
        }

        eventName = 'renamed'

        /**
        * Event with renamed identifier name
        */
        this.$emit(eventName, value)
      },

      /**
      * @protected
      */
      recursiveIdentifierValue () {
        console.log('check')

        let recursiveValue = 'recursive'
        const value = 'event value'

        if (eventName) {
          const x = 'consequent-var-event'

          console.log('>', eventName)
          this.$emit(x, value)
        }

        if (eventName) {
          console.log('>', eventName)
          this.$emit('if-event', value)
        } else if (value) {
          this.$emit('else-if-event', value)
        } else {
          this.$emit('else-event', 123)
        }

        for (let i = 0; i < 0; i++) {
          this.$emit('for-event', value)
        }

        for (let i of []) {
          this.$emit('for-of-event', value)
        }

        for (let i in {}) {
          this.$emit('for-in-event', value)
        }

        do {
          this.$emit('do-while-event', value)
        } while (false)

        while (false) {
          this.$emit('while-event', value)
        }

        switch (x) {
          case 1:
            this.$emit('switch-case-event', value)
            break

          default:
            this.$emit('switch-case-default-event', value)
            break
        }

        try {
          this.$emit('try-event', value)
        } catch (e) {
          this.$emit('try-catch-event', value)
        } finally {
          this.$emit('try-finally-event', value)
        }

        eventName = recursiveValue

        /**
        * Event with recursive identifier name
        */
        this.$emit(eventName, value, 12)
      },

      uncommentedMethod (a, b = 2, c = this.componentName) {},
      withDefault (f = () => 0) {},
      withAlias (a = aliasName) {},
      withSpread (...args) {},
      withDefaultObject (options = {}) {},
      withDestructuring ({ x, y }) {},
    },

    beforeRouteEnter (to, from, next) {
      next((vm) => {
        /**
         * beforeRouteEnter event description
         */
        vm.$emit('beforeRouteEnter-event', this.value)
      })
    },

    beforeRouteUpdate (to, from, next) {
      next((vm) => {
        /**
         * beforeRouteUpdate event description
         */
        vm.$emit('beforeRouteUpdate-event', this.value)
      })
    },

    beforeRouteLeave (to, from, next) {
      next((vm) => {
        /**
         * beforeRouteLeave event description
         */
        vm.$emit('beforeRouteLeave-event', this.value)
      })
    },

    beforeCreate () {
      /**
       * beforeCreate event description
       */
      this.$emit('beforeCreate-event', this.value)
    },

    created () {
      /**
       * Created event description
       */
      this.$emit('created-event', this.value)
    },

    beforeMount () {
      /**
       * beforeMount event description
       */
      this.$emit('beforeMount-event', this.value)
    },

    mounted () {
      /**
       * mounted event description
       */
      this.$emit('mounted-event', this.value)
    },

    beforeUpdate () {
      /**
       * beforeUpdate event description
       */
      this.$emit('beforeUpdate-event', this.value)
    },

    updated () {
      /**
       * updated event description
       */
      this.$emit('updated-event', this.value)
    },

    beforeDestroy () {
      /**
       * beforeDestroy event description
       */
      this.$emit('beforeDestroy-event', this.value)
    },

    destroyed () {
      /**
       * destroyed event description
       */
      this.$emit('destroyed-event', this.value)
    },

    render (createElement, { props }) {
      /**
       * render event description
       */
      this.$emit('render-event', this.value)
    }
  })
</script>
