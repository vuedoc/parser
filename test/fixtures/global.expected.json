{
  "name": "checkboxPointer",
  "description": "The generic component\nSub description",
  "keywords": [
    {
      "name": "public",
      "description": ""
    },
    {
      "name": "alpnum",
      "description": "azert0 123456789"
    },
    {
      "name": "generic",
      "description": "Keyword generic description"
    },
    {
      "name": "multiline",
      "description": "Keyword multiline\ndescription"
    },
    {
      "name": "special-char",
      "description": "{$[ç(àë£€%µù!,|`_\\\\<>/_ç^?;.:/!§)]}"
    },
    {
      "name": "punctuations",
      "description": "!,?;.:!"
    },
    {
      "name": "operators",
      "description": "-/+<>=*%"
    },
    {
      "name": "slot",
      "description": "inputs - Use this slot to define form inputs ontrols"
    },
    {
      "name": "slot",
      "description": "actions - Use this slot to define form action buttons controls"
    },
    {
      "name": "slot",
      "description": "footer - Use this slot to define form footer content."
    },
    {
      "name": "model",
      "description": ""
    }
  ],
  "slots": [
    {
      "kind": "slot",
      "visibility": "public",
      "description": "Default slot",
      "keywords": [],
      "name": "default"
    },
    {
      "kind": "slot",
      "visibility": "public",
      "description": "Use this slot to set the checkbox label",
      "keywords": [],
      "name": "label"
    },
    {
      "kind": "slot",
      "visibility": "public",
      "description": "This\nis multiline description",
      "keywords": [],
      "name": "multiline"
    },
    {
      "kind": "slot",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "undescribed"
    },
    {
      "kind": "slot",
      "visibility": "public",
      "description": "Slot with defined keywords",
      "keywords": [
        {
          "name": "protected",
          "description": ""
        },
        {
          "name": "props",
          "description": "{ name, address, email }"
        }
      ],
      "name": "multiline"
    }
  ],
  "props": [
    {
      "kind": "prop",
      "visibility": "public",
      "description": "The checkbox model",
      "keywords": [
        {
          "name": "model",
          "description": ""
        }
      ],
      "name": "v-model",
      "type": "Array",
      "default": "__undefined__",
      "required": true,
      "describeModel": true
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Initial checkbox state",
      "keywords": [],
      "name": "disabled",
      "type": "Boolean",
      "default": "__undefined__",
      "required": false,
      "describeModel": false
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Initial checkbox value",
      "keywords": [],
      "name": "checked",
      "type": "Boolean",
      "default": true,
      "required": false,
      "describeModel": false
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Prop with arrow function",
      "keywords": [],
      "name": "prop-with-arrow",
      "type": "Object",
      "default": "() => ({ name: 'X'})",
      "required": false,
      "describeModel": false
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Prop with object function",
      "keywords": [],
      "name": "prop-with-function",
      "type": "Object",
      "default": "function() {\n          return { name: 'X' }\n        }",
      "required": false,
      "describeModel": false
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Prop with anonymous function",
      "keywords": [],
      "name": "prop-with-anonymous-function",
      "type": "Object",
      "default": "function() {\n          return { name: 'X' }\n        }",
      "required": false,
      "describeModel": false
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Prop with named function",
      "keywords": [],
      "name": "prop-with-named-function",
      "type": "Object",
      "default": "function propWithNamedFunction() {\n          return { name: 'X' }\n        }",
      "required": false,
      "describeModel": false
    },
    {
      "kind": "prop",
      "visibility": "public",
      "description": "Prop with protected visibility",
      "keywords": [
        {
          "name": "private",
          "description": ""
        }
      ],
      "name": "prop-with-private-visibility",
      "type": "Object",
      "default": "function propWithNamedFunction() {\n          return { name: 'X' }\n        }",
      "required": false,
      "describeModel": false
    }
  ],
  "data": [
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "int",
      "type": "number",
      "initial": 12
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "float",
      "type": "number",
      "initial": 12.2
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "booleanTrue",
      "type": "boolean",
      "initial": true
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "booleanFalse",
      "type": "boolean",
      "initial": false
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "string",
      "type": "string",
      "initial": "Hello"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "string",
      "type": "string",
      "initial": "Hello"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "null",
      "type": "null",
      "initial": null
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "undefined",
      "type": "undefined"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "function",
      "type": "NewExpression",
      "initial": "new Function()"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "arrowFunction",
      "type": "ArrowFunctionExpression"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "date",
      "type": "CallExpression",
      "initial": "Date.now()"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "pointer",
      "type": "string",
      "initial": "pointed value"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "componentName",
      "type": "string",
      "initial": "checkboxPointer"
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "value",
      "type": "null",
      "initial": null
    },
    {
      "kind": "data",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "int",
      "type": "number",
      "initial": 13
    }
  ],
  "computed": [
    {
      "kind": "computed",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "id",
      "dependencies": [
        "componentName",
        "pointer"
      ]
    },
    {
      "kind": "computed",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "name",
      "dependencies": [
        "componentName"
      ]
    },
    {
      "kind": "computed",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "withoutDependency",
      "dependencies": []
    }
  ],
  "events": [
    {
      "kind": "event",
      "visibility": "public",
      "description": "Event with identifier name",
      "keywords": [],
      "name": "check",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "Event with renamed identifier name",
      "keywords": [],
      "name": "renamed",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "consequent-var-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "if-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "for-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "for-of-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "for-in-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "do-while-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "while-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "switch-case-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "switch-case-default-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "try-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "try-catch-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "try-finally-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "Event with recursive identifier name",
      "keywords": [],
      "name": "recursive",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeRouteEnter event description",
      "keywords": [],
      "name": "beforeRouteEnter-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeRouteUpdate event description",
      "keywords": [],
      "name": "beforeRouteUpdate-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeRouteLeave event description",
      "keywords": [],
      "name": "beforeRouteLeave-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeCreate event description",
      "keywords": [],
      "name": "beforeCreate-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "Created event description",
      "keywords": [],
      "name": "created-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeMount event description",
      "keywords": [],
      "name": "beforeMount-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "mounted event description",
      "keywords": [],
      "name": "mounted-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeUpdate event description",
      "keywords": [],
      "name": "beforeUpdate-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "updated event description",
      "keywords": [],
      "name": "updated-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "beforeDestroy event description",
      "keywords": [],
      "name": "beforeDestroy-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "destroyed event description",
      "keywords": [],
      "name": "destroyed-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "render event description",
      "keywords": [],
      "name": "render-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "Template event with @",
      "keywords": [],
      "name": "template-@-event",
      "arguments": []
    },
    {
      "kind": "event",
      "visibility": "public",
      "description": "Template event with v-on",
      "keywords": [],
      "name": "template-v-on-event",
      "arguments": []
    }
  ],
  "methods": [
    {
      "kind": "method",
      "visibility": "private",
      "description": "",
      "keywords": [
        {
          "name": "private",
          "description": ""
        }
      ],
      "name": "privateMethod",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": "Check the checkbox",
      "keywords": [],
      "name": "check",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "protected",
      "description": "",
      "keywords": [
        {
          "name": "protected",
          "description": ""
        }
      ],
      "name": "recursiveIdentifierValue",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "uncommentedMethod",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "withDefault",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "withAlias",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "withSpread",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "withDefaultObject",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    },
    {
      "kind": "method",
      "visibility": "public",
      "description": null,
      "keywords": [],
      "name": "withDestructuring",
      "params": [],
      "return": {
        "type": "void",
        "description": null
      }
    }
  ]
}
